api/v1/tlsroute.go

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2024 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type TargetType string
+
+const (
+	TargetTypeUpstreamTunnel TargetType = "UpstreamTunnel"
+	TargetTypeUI             TargetType = "UI"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Namespaced
+
+type TLSTerminatedRoute struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              TLSTerminatedRouteSpec `json:"spec,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// TLSTerminatedRouteList contains a list of ManagedCluster resources.
+type TLSTerminatedRouteList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []TLSTerminatedRoute `json:"items"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type TLSPassThroughRoute struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Dest is the destination URL
+	Spec TLSPassThroughRouteSpec `json:"spec"`
+}
+
+type TLSPassThroughRouteSpec struct {
+	// +kubebuilder:validation:Enum=UpstreamTunnel
+	// +required
+	Target TargetType `json:"target"`
+
+	// +required
+	SNIMatch *SNIMatch `json:"sniMatch"`
+
+	// Destination is the destination url to proxy the request to.
+	// +required
+	Destination string `json:"destination"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type TLSPassThroughRouteList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []TLSPassThroughRoute `json:"items"`
+}
+
+type TLSTerminatedRouteSpec struct {
+	// +kubebuilder:validation:Enum=UpstreamTunnel;UI
+	Target TargetType `json:"target"`
```

Comment:
> This field name feels wrong - the UI is never the target of these routes, rather it's the source right?

I was thinking more what are the routes for or what server are you targeting, they target the UI or the tunnel which have different servers (at least internally). We could denote this by the source, but the source doesn't necessarily dictate what server is used or how it's routed.

> I also am not sure that the UI is strictly speaking the only possible user of Voltron - other components could route via Voltron, no?

Yes they can, and we have another value for the tunnel. I guess maybe you're getting at that other components could use the same server that the UI is using, so denoting the value as strictly "UI" might not be accurate. I was thinking that this particular service was only used by the UI though, anything in the cluster should not route to backend services via voltron, they should just hit the endpoint directly (with the exception of requests for the tunnel).

Reply:
> But there are many potential backing servers, no?

I think my phrasing was incorrect here, I more meant the proxy server you're intending to hit, what is doing the proxy or what "gateway" (in kubernetes terms) is the route intended for. There are multiple, for the UI, for hte tunnel, and even for the tunnel there's upstream and downstream (is this route to forward traffic going down the tunnel or up from it).

I was thinking that having this target allows it to be flexible and extensible, if we have a new proxy server within voltron (or somewhere else) we can add a new target type.

> There's no such thing as a server in our system called "UI"

The proxy server that serves routes for the UI is what I was thinking of. That's basically a dedicated proxy for the UI (there might be some slight caveats to that, but the fact that there are caveats might be smell tbh).

> Maybe this would be better phrased as Tunnel bool? Basically an indicator for whether this route should be applied to the tunnel or not?

We could, but I'm not sure I see the benefit. Either we take that to mean this route should be configured for the proxy that serves the UI and the upstream traffic from the tunnel (so you can configure both with one) in which case you can't configure routes just for the tunnel, or we need another field to signal this is just for the tunnel and not for the UI proxy. This also makes creating routes for the downstream tunnel more awkward through the API if / when we extend the API to allow this.

So having this "Target" as a string allows this to be more flexible. This is also the terminology that that the kubernetes gateway API uses to  target the what handles proxying the routes. I'm fine to change the name (of both the field an the value UI), but I'd like this to be a string.